src/tools/rbd_mirror/image_map/types.h

```diff
@@ -0,0 +1,161 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#ifndef CEPH_RBD_MIRROR_IMAGE_MAP_TYPES_H
+#define CEPH_RBD_MIRROR_IMAGE_MAP_TYPES_H
+
+#include <set>
+#include <map>
+#include <string>
+#include <memory>
+
+namespace rbd {
+namespace mirror {
+namespace image_map {
+
+enum MapShuffleType {
+  SHUFFLE_TYPE_INSTANCES_ADDED = 0,   // rebalance due to addition of instances
+  SHUFFLE_TYPE_INSTANCES_REMOVED,     // rebalance due to removal of instances
+  SHUFFLE_TYPE_IMAGES_ADDED_REMOVED,  // rebalance due to adding/removing images
+};
+
+struct MappedImage {
+  std::string id;
+  std::string global_id;
+  std::string instance_id;
+
+  MappedImage(const std::string &global_id)
+    : global_id(global_id) {
+  }
+
+  MappedImage(const std::string &id,
+              const std::string &global_id)
+    : id(id), global_id(global_id) {
+  }
+
+  MappedImage(
+    const std::string &id,
+    const std::string &global_id,
+    const std::string &instance_id)
+    : id(id),
+      global_id(global_id),
+      instance_id(instance_id) {
+  }
+
+  inline bool operator==(const MappedImage &rhs) const {
+    return (global_id == rhs.global_id && id == rhs.id);
+  }
+
+  inline bool operator<(const MappedImage &rhs) const {
+    return global_id < rhs.global_id;
+  }
+};
+
+struct RemappedImage {
+  std::string id;
+  std::string global_id;
+  std::string from_instance_id;
+  std::string to_instance_id;
+
+  RemappedImage(const std::string &global_id)
+    : global_id(global_id) {
+  }
+
+  RemappedImage(
+    const std::string &id,
+    const std::string &global_id,
+    const std::string &from_instance_id,
+    const std::string &to_instance_id)
+    : id(id),
+      global_id(global_id),
+      from_instance_id(from_instance_id),
+      to_instance_id(to_instance_id) {
+  }
+
+  inline bool operator==(const RemappedImage &rhs) const {
+    return (global_id == rhs.global_id && id == rhs.id);
+  }
+
+  inline bool operator<(const RemappedImage &rhs) const {
+    return global_id < rhs.global_id;
+  }
+};
+
+std::ostream &operator<<(
+  std::ostream &, const MappedImage &image);
+std::ostream &operator<<(
+  std::ostream &, const RemappedImage &image);
+
+typedef std::set<MappedImage> MappedImages;
+typedef std::set<RemappedImage> RemappedImages;
+
+struct MapSpec {
+  MappedImages   added;
+  MappedImages   removed;
+  RemappedImages remapped;
+
+  MapSpec() {
+  }
+
+  MapSpec(const MappedImages &added,
+          const MappedImages &removed,
+          const RemappedImages &remapped)
+    : added(added),
+      removed(removed),
+      remapped(remapped) {
+  }
+};
+
+std::ostream &operator<<(
+  std::ostream &, const MapSpec &mapspec);
+
+// in-memory map of tracked images
+struct InImage {
```

Comment:
I think you would only need this data structure on a transient basis when loading the map from disk and then the policy's in-memory map would be initialized from it. Of course, the global image id -> instance id map is so simple it could just be represented by a std::map